external-crates/move/crates/move-analyzer/src/completion.rs

```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
+        }
+    }
+
+    let all_identifiers = symbols
+        .file_mods
+        .values()
+        .flatten()
+        .map(|mdef| sp(mdef.name_loc, mdef.ident))
+        .collect::<BTreeSet<_>>();
+    for mod_ident in all_identifiers.into_iter() {
+        if is_pkg_mod_ident(&mod_ident.value, leading_name) {
+            mod_identifiers.insert(mod_ident);
+        }
+    }
+
+    mod_identifiers
+}
+
+/// Computes completions for variants of a given enum
+fn variant_completions(
+    symbols: &Symbols,
+    mod_ident: &ModuleIdent,
+    datatype_name: Symbol,
+) -> Vec<CompletionItem> {
+    let mut completions = vec![];
+    let Some(mod_defs) = symbols
+        .file_mods
+        .values()
+        .flatten()
+        .find(|mdef| mdef.ident == mod_ident.value)
+    else {
+        return completions;
+    };
+
+    let Some(edef) = mod_defs.enums.get(&datatype_name) else {
+        return completions;
+    };
+
+    let Some(DefInfo::Enum(.., variants, _)) = symbols.def_info.get(&edef.name_loc) else {
+        return completions;
+    };
+
+    for vinfo in variants {
+        let Some(DefInfo::Variant(_, _, vname, is_positional, field_names, ..)) =
+            symbols.def_info.get(&vinfo.name.loc)
+        else {
+            continue;
+        };
+
+        let label = if field_names.is_empty() {
+            vname.to_string()
+        } else if *is_positional {
+            format!("{vname}()")
+        } else {
+            format!("{vname}{{}}")
+        };
+        let field_snippet = field_names
+            .iter()
+            .enumerate()
+            .map(|(snippet_idx, fname)| {
+                if *is_positional {
+                    format!("${{{}}}", snippet_idx + 1)
+                } else {
+                    format!("${{{}:{}}}", snippet_idx + 1, fname)
+                }
+            })
+            .collect::<Vec<_>>()
+            .join(", ");
+        let insert_text = if *is_positional {
+            format!("{vname}({field_snippet})")
+        } else {
+            format!("{vname}{{{field_snippet}}}")
+        };
+
+        completions.push(CompletionItem {
+            label,
+            kind: Some(CompletionItemKind::ENUM_MEMBER),
+            insert_text: Some(insert_text),
+            insert_text_format: Some(InsertTextFormat::SNIPPET),
+            ..Default::default()
+        });
+    }
+
+    completions
+}
+
+/// Computes completions for access path entries (components beyond the leading name)
+fn entries_completions(
+    symbols: &Symbols,
+    cursor: &CursorContext,
+    info: &AliasAutocompleteInfo,
+    prev_info: ChainComponentInfo,
+    chain_target: ChainCompletionTarget,
+    path_entries: &[Name],
+    path_index: usize,
+    colon_colon_triggered: bool,
+    completions: &mut Vec<CompletionItem>,
+) {
+    let ChainComponentInfo {
+        loc: prev_loc,
+        kind: prev_kind,
+    } = prev_info;
+
+    let mut at_colon_colon = false;
```

Comment:
Please split this up into a few smaller functions, maybe something like this:

```suggestion
/// Walks down a name chain, looking for the relevant portion that contains the cursor. When. it finds, it calls to `name_chain_entry_completions` to compute and return the completions.
fn completions_for_name_chain_entry(
    symbols: &Symbols,
    cursor: &CursorContext,
    info: &AliasAutocompleteInfo,
    prev_info: ChainComponentInfo,
    chain_target: ChainCompletionTarget,
    path_entries: &[Name],
    path_index: usize,
    colon_colon_triggered: bool,
) -> Vec<CompletionItem> {
    let ChainComponentInfo {
        loc: prev_loc,
        kind: prev_kind,
    } = prev_info;

    let at_colon_colon =  cursor_at_colon_colon(path_entries, path_index, cursor);
    // we are at `::`, or at some component's identifier
    if at_colon_colon || path_entries[path_index].loc.contains(&cursor.loc) {
        name_chain_entry_completions(
            symbols, cursor, info, ..., prev_kind
        )
    } else if let Some(next_kind) = next_name_chain_kind(..., prev_kind) {
            entries_completions(
                symbols,
                cursor,
                info,
                ChainComponentInfo::new(component_name.loc, next_kind),
                chain_target,
                path_entries,
                path_index + 1,
                colon_colon_triggered,
            )
    } else {
        vec![]
    }
```

You could also consider doing it in two steps, so replace the call to `entries_completions` with:

```
if let Some(completion_context_info) = find_completion_info(symbols, cursor, ...) {
  completions.extend(name_chain_entry_completions(symbols, ..., completion_context_info));
}
```